pkg/cluster/manager.go

```diff
@@ -0,0 +1,1886 @@
+
+	return nil
+}
+
+// ScaleOut scale out the cluster.
+func (d *Manager) ScaleOut(
+	clusterName string,
+	topoFile string,
+	afterDeploy func(b *task.Builder, newPart spec.Topology),
+	final func(b *task.Builder, name string, meta spec.Metadata),
+	opt ScaleOutOptions,
+	skipConfirm bool,
+	optTimeout int64,
+	sshTimeout int64,
+	nativeSSH bool,
+) error {
+	metadata, err := d.meta(clusterName)
+	if err != nil {
+		return perrs.AddStack(err)
+	}
+
+	topo := metadata.GetTopology()
+	base := metadata.GetBaseMeta()
+
+	// not allowing validation errors
+	if err := topo.Validate(); err != nil {
+		return err
+	}
+
+	// Inherit existing global configuration. We must assign the inherited values before unmarshalling
+	// because some default value rely on the global options and monitored options.
+	newPart := topo.NewPart()
+
+	// The no tispark master error is ignored, as if the tispark master is removed from the topology
+	// file for some reason (manual edit, for example), it is still possible to scale-out it to make
+	// the whole topology back to normal state.
+	if err := clusterutil.ParseTopologyYaml(topoFile, newPart); err != nil &&
+		!errors.Is(perrs.Cause(err), spec.ErrNoTiSparkMaster) {
+		return err
+	}
+
+	if err := validateNewTopo(newPart); err != nil {
+		return err
+	}
+
+	// Abort scale out operation if the merged topology is invalid
+	mergedTopo := topo.MergeTopo(newPart)
+	if err := mergedTopo.Validate(); err != nil {
+		return err
+	}
+
+	if err := prepare.CheckClusterPortConflict(d.specManager, clusterName, mergedTopo); err != nil {
+		return err
+	}
+	if err := prepare.CheckClusterDirConflict(d.specManager, clusterName, mergedTopo); err != nil {
+		return err
+	}
+
+	patchedComponents := set.NewStringSet()
+	newPart.IterInstance(func(instance spec.Instance) {
+		if utils.IsExist(d.specManager.Path(clusterName, spec.PatchDirName, instance.ComponentName()+".tar.gz")) {
+			patchedComponents.Insert(instance.ComponentName())
+		}
+	})
+
+	if !skipConfirm {
+		// patchedComponents are components that have been patched and overwrited
+		if err := d.confirmTopology(clusterName, base.Version, newPart, patchedComponents); err != nil {
+			return err
+		}
+	}
+
+	sshConnProps, err := cliutil.ReadIdentityFileOrPassword(opt.IdentityFile, opt.UsePassword)
+	if err != nil {
+		return err
+	}
+
+	// Build the scale out tasks
+	t, err := buildScaleOutTask(d, clusterName, metadata, mergedTopo, opt, sshConnProps, newPart, patchedComponents, optTimeout, sshTimeout, nativeSSH, afterDeploy, final)
+	if err != nil {
+		return err
+	}
+
+	if err := t.Execute(task.NewContext()); err != nil {
+		if errorx.Cast(err) != nil {
+			// FIXME: Map possible task errors and give suggestions.
+			return err
+		}
+		return perrs.Trace(err)
+	}
+
+	log.Infof("Scaled cluster `%s` out successfully", clusterName)
+
+	return nil
+}
+
+func (d *Manager) meta(name string) (metadata spec.Metadata, err error) {
```

Comment:
Prefer to rename all method's receiver of `*Manager`.

Reply:
resolve by https://github.com/pingcap/tiup/pull/587/commits/06db827e3b1d4e01a2bc61cafa09a15561a0db58